Add support for programmatic title generation

NeevCohen opened the pull request:
<!-- Thank you for your contribution! -->
<!-- Unless your change is trivial, please create an issue to discuss the change before creating a PR -->

## Change Summary

Implement programmatic title generation via `ConfigDict` and `Field`

<!-- Please give a short summary of the changes. -->

## Related issue number

fix #4632 

<!-- WARNING: please use "fix #123" style references so the issue is closed when this PR is merged. -->

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [x] Tests pass on CI
* [x] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**


Selected Reviewer: @dmontagu

NeevCohen commented:
Please review

sydney-runkle commented on pydantic/fields.py:
```diff
@@ -213,6 +221,8 @@ def __init__(self, **kwargs: Unpack[_FieldInfoInputs]) -> None:
         self.serialization_alias = kwargs.pop('serialization_alias', None)
         alias_is_set = any(alias is not None for alias in (self.alias, self.validation_alias, self.serialization_alias))
         self.alias_priority = kwargs.pop('alias_priority', None) or 2 if alias_is_set else None
+        self.field_title_generator = kwargs.pop('field_title_generator', None)
+        self.title_priority = kwargs.pop('title_priority', None) or 2 if self.title is not None else None
```
Very minor nit - can we order these in the same order as seen in the attribute docstring?

sydney-runkle commented on pydantic/_internal/_generate_schema.py:
```diff
@@ -1105,6 +1151,10 @@ def set_discriminator(schema: CoreSchema) -> CoreSchema:
         schema = self._apply_field_serializers(
             schema, filter_field_decorator_info_by_field(decorators.field_serializers.values(), name)
         )
+        field_title_generator = field_info.field_title_generator or self._config_wrapper.field_title_generator
+        if field_title_generator is not None:
+            self._apply_field_title_generator_to_field_info(field_title_generator, field_info, name)
+
```
Looks like we repeat this exact same logic 3x at least, maybe use a function to abstract away some of this?

sydney-runkle commented on pydantic/_internal/_generate_schema.py:
```diff
@@ -202,19 +201,26 @@ def apply_each_item_validators(
         schema_or_field: The schema data to generate a JSON schema from.
         handler: The `GetCoreSchemaHandler` instance.
         cls: The model-like class.
+        title: The title to set for the model's schema, defaults to the models name
```
```suggestion
        title: The title to set for the model's schema, defaults to the model's name
```

sydney-runkle reviewed (changes requested):
Hey @NeevCohen,

Great work, as usual!

I want to look again at `_generate_schema.py` and do a more in-depth look at your tests (great work btw, looks very thorough). This looks very promising though - I've included some initial feedback / some minor change requests.

One other change I'd like to see is the addition of some documentation in the `concepts` section of our docs!

NeevCohen commented:
@sydney-runkle Thanks! I implemented the changes you requested. 
Just one question, in what part of the `concepts` section do you think I should document this feature in?

PrettyWood reviewed (commented):
Question : why not having title be a string or a callable that returns a string instead of two params?

NeevCohen commented:
@PrettyWood I implemented it the way the original issue suggested, I'm fine with changing it if we decide that is the way we want to go. 
Also, since this is similar to the `alias` and `alias_generator` parameters, I figure it would be better to consistent with the API. IMO if we change `title` to be a string or a callable, we should also change `alias` to be the same way.

PrettyWood commented:
@NeevCohen totally agree. Probably a change for v3 then

NeevCohen commented:
@PrettyWood Awesome.
@sydney-runkle did you get a chance to do a more in-depth look?

sydney-runkle commented on docs/concepts/json_schema.md:
```diff
@@ -1029,6 +1076,94 @@ print(json.dumps(TypeAdapter(Person).json_schema(), indent=2))
+
+
+class Person(BaseModel):
+    model_config = ConfigDict(field_title_generator=to_pascal)
```
Maybe we could use a different generator here - `to_pascal` results in the same names as the default capitalization for `name` and `age`

sydney-runkle replied:
Could either go with `upper` or something like your example in the class title case below :)

sydney-runkle commented on pydantic/fields.py:
```diff
@@ -977,6 +994,7 @@ class ComputedFieldInfo:
         alias: The alias of the property to be used during serialization.
         alias_priority: The priority of the alias. This affects whether an alias generator is used.
         title: Title of the computed field to include in the serialization JSON schema.
+        title_priority: Priority of the title. This affects whether a title generator is used.
```
```suggestion
        title_priority: Priority of the title. This affects whether a title generator is used.
        field_title_generator: A callable that takes a field name and returns title for it.
```

sydney-runkle commented on pydantic/_internal/_generate_schema.py:
```diff
@@ -225,10 +231,12 @@ def modify_model_json_schema(
         original_schema['title'] = cls.__name__
-    # BaseModel; don't use cls.__doc__ as it will contain the verbose class signature by default
-    docstring = None if cls is BaseModel else cls.__doc__
+    # BaseModel + Dataclass; don't use cls.__doc__ as it will contain the verbose class signature by default
```
Nice!

sydney-runkle reviewed (approved):
@NeevCohen,

Great work here. Really impressed with the thorough testing, and great work on the docs as well.

sydney-runkle commented:
Let's have @samuelcolvin do a final review of this given that he made the original feature request. Stay tuned!

samuelcolvin commented on docs/concepts/json_schema.md:
```diff
@@ -490,6 +535,8 @@ Specifically, the following config options are relevant:
 * [`schema_generator`][pydantic.config.ConfigDict.schema_generator]
 * [`json_schema_mode_override`][pydantic.config.ConfigDict.json_schema_mode_override]
+* [`field_title_generator`][pydantic.config.ConfigDict.field_title_generator]
+* [`class_title_generator`][pydantic.config.ConfigDict.class_title_generator]
```
I would suggest `model_title_generator` might be a better name.

samuelcolvin commented on docs/concepts/json_schema.md:
````diff
@@ -1029,6 +1076,93 @@ print(json.dumps(TypeAdapter(Person).json_schema(), indent=2))
 ```
 
 
+### Using `field_title_generator`
````
I'm not sure why we have two sections of docs about one argument?

NeevCohen replied:
Once for the model level and once for the model level.

samuelcolvin commented on docs/concepts/json_schema.md:
```diff
@@ -1029,6 +1076,93 @@ print(json.dumps(TypeAdapter(Person).json_schema(), indent=2))
+
+
+class Person(BaseModel):
+    model_config = ConfigDict(field_title_generator=lambda m: m.upper())
```
```suggestion
    model_config = ConfigDict(field_title_generator=str.upper)
```

surely? (unless we pass other argument as I suggested above)

samuelcolvin commented on docs/concepts/json_schema.md:
```diff
@@ -1029,6 +1076,93 @@ print(json.dumps(TypeAdapter(Person).json_schema(), indent=2))
+from pydantic import BaseModel, ConfigDict
+
+
+def make_title(field_name: str) -> str:
```
again I would suggest we pass the `cls` itself, as well as the current name.

NeevCohen replied:
Sure. 
What type do you reckon the `cls` parameter be annotated as? `Type[BaseModel]`?

samuelcolvin commented on pydantic/config.py:
```diff
@@ -33,11 +33,18 @@ class ConfigDict(TypedDict, total=False):
     """The title for the generated JSON schema, defaults to the model's name"""
 
+    class_title_generator: Callable[[str], str] | None
+    """A callable that takes a class name and returns the title for it. Defaults to `None`."""
```
we should be clear about whether this works for:
* pydantic models?
* dataclasses?
* typeddicts?

NeevCohen replied:
I works for all model and model-like (pydantic dataclass, builtin dataclass, typeddict) types. `ConfigDict` can be applied to all of them so IMO it's a little redundant. Changing the name to `model_title_generator` makes this even more clear I feel.


I will add more documentation if you feel it's necessary.

samuelcolvin commented on pydantic/fields.py:
```diff
@@ -106,6 +108,8 @@ class FieldInfo(_repr.Representation):
         validation_alias: The validation alias of the field.
         serialization_alias: The serialization alias of the field.
         title: The title of the field.
+        title_priority: Priority of the field's title. This affects whether a title generator is used.
```
Are you sure we need this? If so, I think it needs more documentation.

NeevCohen replied:
I guess I wanted to have a similar API to `alias_generator` and `alias_priority`.

sydney-runkle commented:
@NeevCohen,

Sorry for the delay here, we'll get this across the line for 2.8!

samuelcolvin reviewed (commented):
overall looks like a good idea, but there are a few things to iron out.

NeevCohen commented on docs/concepts/json_schema.md:
```diff
@@ -481,6 +482,50 @@ print(json.dumps(Foo.model_json_schema(), indent=2))
+from pydantic import BaseModel, Field
+
+
+def make_title(field_name: str) -> str:
```
The full field as in the `FieldInfo` object? That we need to pass both `FieldInfo` and the name of the field since `FieldInfo` doesn't include the name.

Then the signature of `make_title` would be 

```python
def make_title(field_name: str, field_info: FieldInfo) -> str: ...
```

sydney-runkle commented:
@NeevCoehn,

Just a side note, we have a `pydantic` slack that you can join [here](https://join.slack.com/t/pydanticlogfire/shared_invite/zt-2b57ljub4-936siSpHANKxoY4dna7qng) for faster iteration on development questions, etc :).

Thanks for all of your great work!

NeevCohen commented:
@sydney-runkle Awesome! Will do!

Thanks

sydney-runkle commented:
@NeevCohen,

I think @samuelcolvin is quite busy this week - I'll take another pass + review this today. Great work thus far! I think we can definitely get this across the line soon 🚀.

sydney-runkle commented:
> Add more documentation for the title_priority logic

@NeevCohen, lmk when you've done this and I can do a final review!

> Decide if the title generator function should take an argument of type FieldInfo as well

I think this could be useful down the line, but it's not a breaking change to introduce later bc we can do signature introspection, so I don't think we need to include this in this PR.

> Just want to circle back - are we ok with adding the `title` argument to the `modify_model_json schema` function? In general, I'm not super happy with the way we prep to modify JSON schema while building the core schema, so I'm not super excited about this change...

Yeah, the change I'm requesting above is major, I think this is ok to add for now!

NeevCohen commented:
@sydney-runkle regarding `title_priority`, after giving it some more thought I feel like it is not necessary. When I first implemented it, I did what `alias_generator` did. `title_priority` can be used to determine the priority of `title` over `title_generator`, but I think that specifying a `title` should always take precedence over `title_generator`, and there is no real use-case for `title_priority`, so I think I'll just remove it all together.

Regarding adding another `FieldInfo` parameter to the `title_generator`, I think that's a good idea and I can see how someone would want to generate a title based on more than just the name of the field, so I think I'll add it in this PR.

sydney-runkle commented:
> @sydney-runkle regarding title_priority, after giving it some more thought I feel like it is not necessary. When I first implemented it, I did what alias_generator did. title_priority can be used to determine the priority of title over title_generator, but I think that specifying a title should always take precedence over title_generator, and there is no real use-case for title_priority, so I think I'll just remove it all together.

Go for it. The more simple, the better. We can always add that down the line if some realistic use case is presented.

> Regarding adding another FieldInfo parameter to the title_generator, I think that's a good idea and I can see how someone would want to generate a title based on more than just the name of the field, so I think I'll add it in this PR.

Sure thing. Ping me when you'd like a review, sounds great!

sydney-runkle commented:
@NeevCohen,

Hoping to release v2.8 before the end of June - any chance you'll have those updates in by then? Thanks for all of your work on this awesome new feature!

NeevCohen commented:
Hey @sydney-runkle, sorry for the delay, I've been a little busy at work 😅 
Anyway, I've implemented the changed you've requested and some more:

1. I've removed `title_priority` from `FieldInfo`
2. I've changed `field_title_generator` to accept the info of the field as well
3. I've changed `model_title_generator` to accept the class itself, instead of just its name

sydney-runkle commented on pydantic/_internal/_generate_schema.py:
```diff
@@ -1523,6 +1578,7 @@ def _dataclass_schema(
                         dataclass_bases_stack.enter_context(self._types_namespace_stack.push(dataclass_base))
 
                 # Pushing a config overwrites the previous config, so iterate though the MRO backwards
+                config = None
```
I have a suspicion that this isn't needed, but perhaps that's wrong

sydney-runkle reviewed (approved):
This is truly impressive. Great work on this significant feature!!

Thanks for your amazing work on this @NeevCohen!

sydney-runkle commented:
If someone would like to add some more robust testing showcasing how to use both the field name and field info in the title generator function, that would be awesome! (great first issue)
